File: packages/ledger/src/streamConsumer.ts
import type Redis from 'ioredis';
import { deserializeEvent, type EngineEvent } from '@exchange/common';

/**
 * Consume the engine.events stream. Yields events in two phases:
 *
 *   1. REPLAY: every event currently in the stream, oldest first.
 *      We do this with XRANGE so it's a clean drain, not a blocking read.
 *   2. LIVE:   subsequent events as they are appended.
 *      We do this with XREAD BLOCK 0.
 *
 * The consumer maintains a cursor (the last stream id processed) so the
 * transition between phases is seamless and we never miss an event nor
 * see one twice in the same run.
 *
 * Each yielded value carries the mode so the caller can apply different
 * settlement logic during replay vs live (see Settler).
 */
export interface StreamItem {
  event: EngineEvent;
  mode: 'replay' | 'live';
}

const PAGE = 1000;

export async function* consumeEngineStream(
  redis: Redis,
  streamKey: string = 'engine.events'
): AsyncGenerator<StreamItem> {
  // --- REPLAY phase ---
  let cursor: string = '-';
  while (true) {
    const result = (await redis.xrange(
      streamKey,
      cursor,
      '+',
      'COUNT',
      PAGE
    )) as [string, string[]][];
    if (!result || result.length === 0) break;
    for (const [id, fields] of result) {
      const ev = decodeFields(fields);
      if (ev) yield { event: ev, mode: 'replay' };
      cursor = `(${id}`;
    }
    if (result.length < PAGE) break;
  }

  // After replay, `cursor` is the exclusive next id ("(<id>"). For XREAD
  // we need the inclusive last id we saw. Strip the leading `(`. If no
  // events were drained, start from '0' so XREAD streams everything new
  // from the start of the next push.
  let lastId: string = cursor.startsWith('(') ? cursor.slice(1) : '0';

  // --- LIVE phase ---
  while (true) {
    const reply = (await redis.call(
      'XREAD',
      'BLOCK',
      '0',
      'COUNT',
      String(PAGE),
      'STREAMS',
      streamKey,
      lastId
    )) as [string, [string, string[]][]][] | null;
    if (!reply) continue;
    for (const [, entries] of reply) {
      for (const [id, fields] of entries) {
        const ev = decodeFields(fields);
        if (ev) yield { event: ev, mode: 'live' };
        lastId = id;
      }
    }
  }
}

function decodeFields(fields: string[]): EngineEvent | null {
  const idx = fields.indexOf('data');
  if (idx < 0 || idx + 1 >= fields.length) return null;
  return deserializeEvent(fields[idx + 1]!);
}

File: packages-go/common/redis/stream_consumer.go
package redis

import (
	"context"
	"strings"

	"github.com/KaustubhPatange/exchange/common"
	"github.com/redis/go-redis/v9"
)

const DefaultStreamKey = "engine.events"

type StreamMode string

const (
	StreamModeReplay StreamMode = "replay"
	StreamModeLive   StreamMode = "live"
)

type StreamItem struct {
	Event *common.EngineEvent
	Mode  StreamMode
}

func consumeEngineStream(ctx context.Context, client *redis.Client) (<-chan StreamItem, <-chan error) {
	events := make(chan StreamItem)
	errs := make(chan error, 1)

	go func() {
		defer close(events)
		defer close(errs)

		from := "-"
		page := int64(1000)

		// replay mode
		for {
			result, err := client.XRangeN(ctx, DefaultStreamKey, from, "+", page).Result()
			if err != nil {
				errs <- err
				break
			}
			if len(result) == 0 {
				break
			}

			for _, entry := range result {
				event, err := decodeFields(entry)
				if err != nil {
					errs <- err
					break
				}

				select {
				case events <- StreamItem{Event: &event, Mode: StreamModeReplay}:
				case <-ctx.Done():
					return
				}
				from = "(" + entry.ID
			}

			if int64(len(result)) < page {
				break
			}
		}

		lastID := "0"
		if strings.HasPrefix(from, "(") {
			lastID = from[1:]
		}

		// live mode
		for {
			result, err := client.XRead(ctx, &redis.XReadArgs{
				Streams: []string{DefaultStreamKey, lastID},
				Block:   0,
				Count:   page,
			}).Result()
			if err != nil {
				if ctx.Err() != nil {
					return
				}
				errs <- err
				continue
			}

			for _, stream := range result {
				for _, entry := range stream.Messages {
					event, err := decodeFields(entry)
					if err != nil {
						errs <- err
						continue
					}

					select {
					case events <- StreamItem{Event: &event, Mode: StreamModeLive}:
					case <-ctx.Done():
						return
					}
					lastID = entry.ID
				}
			}
		}
	}()

	return events, errs
}

func decodeFields(entry redis.XMessage) (common.EngineEvent, error) {
	data, ok := entry.Values["data"]
	if !ok {
		return nil, nil
	}

	payload, ok := data.(string)
	if !ok {
		return nil, nil
	}

	return common.DeserializeEvent(payload)
}
